src/audio/xm64.c

```diff
@@ -0,0 +1,240 @@
+
+			// Play the waveform. Notice that the waveform might already
+			// being played in this channel, in which case the play
+			// command only resets its position to 0, and keep the sample
+			// buffer full, which is what we want.
+			// The mixer doesn't currently allow for mixer_ch_play() to keep
+			// the current position, but even if it did, xm_tick() might
+			// have changed it since last tick, because there is a XM effect
+			// to force the position in the sample. So it's better to
+			// set it every time with mixer_ch_set_pos.
+			mixer_ch_play(first_ch+i, w);
+			mixer_ch_set_pos(first_ch+i, ch->sample_position);
+
+			// Configure also frequency and volume that might have changed
+			// since last tick.
+			// debugf("vol %f %f\n", ch->actual_volume, ch->actual_panning);
+			mixer_ch_set_freq(first_ch+i, ch->frequency);
+			mixer_ch_set_vol(first_ch+i,
+				muted ? 0 : gvol * ch->actual_volume[0],
+				muted ? 0 : gvol * ch->actual_volume[1]);
+		} else {
+			// No sample in this channel: the channel is mute. Just stop it.
+			mixer_ch_stop(first_ch+i);
+		}
+	}
+
+	// Schedule next tick according to the number of samples in this tick.
+	int delay = ceilf(ctx->remaining_samples_in_tick);
+	ctx->remaining_samples_in_tick -= delay;
+	return delay;
+}
+
+void xm64player_open(xm64player_t *player, const char *fn) {
+	memset(player, 0, sizeof(*player));
+
+	// No pending seek at the moment, we start from beginning anyway.
+	player->seek.patidx = -1;
+
+	char romfn[strlen(fn)+5+1];
+	snprintf(romfn, sizeof(romfn), "rom:/%s", fn);
+	player->fh = fopen(romfn, "rb");
+	assertf(player->fh, "Cannot open file: %s", fn);
+
+	// Load the XM context
+	int sample_rate = audio_get_frequency();
+	assertf(sample_rate >= 0, "audio_init() and mixer_init() must be called before xm64player_open()");
+	if (xm_context_load(&player->ctx, player->fh, sample_rate) != 0) {
+		// Check if the file looks like a standard XM, so to provide
+		// a clear message in that case.
+		char signature[16] = {0};
+		fseek(player->fh, 0, SEEK_SET);
+		fread(signature, 1, 15, player->fh);
+		if (strcmp(signature, "Extended Module") == 0) {
+			assertf(0, "cannot load XM file %s -- please convert to XM64 with audio64", fn);
+		}
+		assertf(0, "error loading XM64 file -- file corrupted?");
+	}
+
+	uint32_t base_rom_addr = dfs_rom_addr(fn);
+
+	// Count samples
+	int ninst = xm_get_number_of_instruments(player->ctx);
+	int nwaves = 0;
+	for (int i=0;i<ninst;i++)
+		nwaves += xm_get_number_of_samples(player->ctx, i+1);
+
+	// Allocate waveforms (one per XM64's "samples" aka waveforms)
+	player->waves = malloc(sizeof(waveform_t) * nwaves);
+	assert(player->waves);
+	int nw = 0;
+	for (int i=0;i<ninst;i++) {
+		xm_instrument_t *inst = &player->ctx->module.instruments[i];
+		for (int j=0;j<inst->num_samples;j++) {
+			xm_sample_t *samp = &inst->samples[j];
+
+			// Convert offset of samples from relative to absolute
+			samp->data8_offset += base_rom_addr;
```

Comment:
The quoted code converts the waveform offset from being file-relative to absolute ROM address. Currently `xm64player` only supports ROM, so if we switch to prefixed paths, I would have to assert on paths which are not `rom:/`. In the future it might be possible to add support to other filesystems like SDFS, but they need to be fast enough for audio not to crack. I think we will have to try and see if they work.

Reply:
Done